core/services/ocr2/plugins/mercury/plugin.go

```diff
@@ -31,42 +37,93 @@ func NewServices(
 	argsNoPlugin libocr2.MercuryOracleArgs,
 	cfg Config,
 	chEnhancedTelem chan ocrcommon.EnhancedTelemetryMercuryData,
-	chainHeadTracker mercury.ChainHeadTracker,
-	orm mercury.DataSourceORM,
+	chainHeadTracker types.ChainHeadTracker,
+	orm types.DataSourceORM,
 	feedID [32]byte,
 ) ([]job.ServiceCtx, error) {
 	if jb.PipelineSpec == nil {
 		return nil, errors.New("expected job to have a non-nil PipelineSpec")
 	}
+
 	var pluginConfig config.PluginConfig
 	err := json.Unmarshal(jb.OCR2OracleSpec.PluginConfig.Bytes(), &pluginConfig)
 	if err != nil {
 		return nil, errors.WithStack(err)
 	}
-	err = config.ValidatePluginConfig(pluginConfig)
+	err = config.ValidatePluginConfig(pluginConfig, feedID)
 	if err != nil {
 		return nil, err
 	}
 	lggr = lggr.Named("MercuryPlugin").With("jobID", jb.ID, "jobName", jb.Name.ValueOrZero())
-	ds := mercury.NewDataSource(
-		orm,
-		pipelineRunner,
-		jb,
-		*jb.PipelineSpec,
-		lggr,
-		runResults,
-		chEnhancedTelem,
-		chainHeadTracker,
-		ocr2Provider.ContractTransmitter(),
-		pluginConfig.InitialBlockNumber.Ptr(),
-		feedID,
-	)
-	argsNoPlugin.MercuryPluginFactory = relaymercury.NewFactory(
-		ds,
-		lggr,
-		ocr2Provider.OnchainConfigCodec(),
-		ocr2Provider.ReportCodec(),
-	)
+
+	version, err := decoder.DecodeSchemaVersionFromFeedId(feedID)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	switch version {
+	case 1:
+		ds := mercuryv1.NewDataSource(
+			orm,
+			pipelineRunner,
+			jb,
+			*jb.PipelineSpec,
+			lggr,
+			runResults,
+			chEnhancedTelem,
+			chainHeadTracker,
+			ocr2Provider.ContractTransmitter(),
+			pluginConfig.InitialBlockNumber.Ptr(),
+			feedID,
+		)
+		argsNoPlugin.MercuryPluginFactory = relaymercuryv1.NewFactory(
+			ds,
+			lggr,
+			ocr2Provider.OnchainConfigCodec(),
+			ocr2Provider.ReportCodecV1(),
+		)
+	case 2:
+		ds := mercuryv2.NewDataSource(
+			pipelineRunner,
+			jb,
+			*jb.PipelineSpec,
+			feedID,
+			lggr,
+			runResults,
+			chEnhancedTelem,
+			ocr2Provider.ContractTransmitter(),
+			types.FeedID(*pluginConfig.LinkFeedID),
+			types.FeedID(*pluginConfig.NativeFeedID),
+		)
+		argsNoPlugin.MercuryPluginFactory = relaymercuryv2.NewFactory(
+			ds,
+			lggr,
+			ocr2Provider.OnchainConfigCodec(),
+			ocr2Provider.ReportCodecV2(),
+		)
+	case 3:
+		ds := mercuryv3.NewDataSource(
+			pipelineRunner,
+			jb,
+			*jb.PipelineSpec,
+			feedID,
+			lggr,
+			runResults,
+			chEnhancedTelem,
+			ocr2Provider.ContractTransmitter(),
+			types.FeedID(*pluginConfig.LinkFeedID),
+			types.FeedID(*pluginConfig.NativeFeedID),
+		)
+		argsNoPlugin.MercuryPluginFactory = relaymercuryv3.NewFactory(
+			ds,
+			lggr,
+			ocr2Provider.OnchainConfigCodec(),
+			ocr2Provider.ReportCodecV3(),
+		)
```

Comment:
2 and 3 look identical, can they be combined?

Reply:
The report codec is different